### PR TITLE
Datasource: Fix stable sort order of query responses

### LINKED ISSUE #41303: Graph panel(old): Order of timeseries under Version 8.2.3 is not working
**What happened**:
It is happened like in forum: https://community.grafana.com/t/order-of-stacked-timeseries-and-legend/21662/3

The order of timeseries under Version 8.2.3 is not ordered like order of queries in query list:
140282685-9c85f4a7-c4f9-4c19-9c1f-d74cf0d2f478

After refreshing panel:
140282726-dd8a2aac-f598-476f-bce4-b2d5c234bb3f

**What you expected to happen**:
It should be ordered like in query list (F1,F2,....) (screenshot):

Unbenannt

**How to reproduce it**:
- Create a dashboard
- Create a old graph panel.
- Input any query with CloudWatcher Metrics related to your data into the query panel
- Refresh the old graph panel to test the order of timeseries

**Anything else we need to know?**:
- The order of timeseries under version 7 was correct.

**Environment**:
- Grafana version: 8.2.3

**Misc**:
Please note:

- all orders of timeseries have suffered this issue when updating from 7.x to 8.2.3
- no configuration has changed
- grafana still displays data.
- we are still using old graph panel

Thank you very much for your help!

### LINKED ISSUE #41773: Inconsistency between graph (stack) visualisation
**What happened**:
In grafana 8.2.3 by having "graph (old)" (same happen even by changing it to "timeseries"), we have inconsistent visualisation of the values between each refresh while the stack mode is on. In the following gif you can observe that each time i'm pressing refresh, we have completely different visualisation. just in one case it's the one we expect, the rest are misleading.

The source of the values are CloudWatch metrics.

grafana-issue

**What you expected to happen**:
This shot is what i expected to have persistently and seems correct:

**Anything else we need to know?**:
Thought be related to [this Description](https://github.com/grafana/grafana/issues/9705#issuecomment-349438920), but changing the interval time and period didn't improved the situation.

**Environment**:
- Grafana version: 8.2.3
- Data source type & version: CloudWatch 5.0.0

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Using changes in https://github.com/grafana/grafana-plugin-sdk-go/pull/425. Been testing this using chrome developer tools and with these changes query responses are always coming back ordered, e.g. A, C given query C, A.

**Which issue(s) this PR fixes**:
Fixes #41773
Fixes #41303
Ref https://github.com/grafana/grafana-plugin-sdk-go/issues/366
Ref #39614

**Special notes for your reviewer**:
- Not sure this fixes all problems for CloudWatch - might need client side changes as well perhaps, see below. I've only tested with multiple CloudWatch queries that returns single series and that seems to work as expected. Before changes testing this I could see color changing between series, but haven't been able to reproduce this with these changes at least
- Is there some additional problem besides ordering of query responses (refIDs), like the order of frames?
- Make sure to kill `make run` and start `make run` when checking out this code - I had some issues not getting the SDK changes before doing this.

Used this dashboard with the test data source, but it seems to retain series, legend and colors both before and after my changes which makes me believe that CloudWatch might do something client side which creates issues.
```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 3340,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "C",
          "scenarioId": "random_walk",
          "seriesCount": 1
        },
        {
          "alias": "",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "hide": false,
          "refId": "A",
          "scenarioId": "random_walk",
          "seriesCount": 1
        }
      ],
      "title": "Time series (queries with single series)",
      "type": "timeseries"
    },
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 3,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "C",
          "scenarioId": "random_walk",
          "seriesCount": 3
        },
        {
          "alias": "",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "hide": false,
          "refId": "A",
          "scenarioId": "random_walk",
          "seriesCount": 3
        }
      ],
      "title": "Time series (queries with multiple series)",
      "type": "timeseries"
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 9
      },
      "hiddenSeries": false,
      "id": 4,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "8.3.0-pre",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "C",
          "scenarioId": "random_walk",
          "seriesCount": 1
        },
        {
          "alias": "",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "hide": false,
          "refId": "A",
          "scenarioId": "random_walk",
          "seriesCount": 1
        }
      ],
      "thresholds": [],
      "timeRegions": [],
      "title": "Graph old (queries with single series)",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "mode": "time",
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "logBase": 1,
          "show": true
        },
        {
          "format": "short",
          "logBase": 1,
          "show": true
        }
      ],
      "yaxis": {
        "align": false
      }
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 12,
        "y": 9
      },
      "hiddenSeries": false,
      "id": 5,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "8.3.0-pre",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "C",
          "scenarioId": "random_walk",
          "seriesCount": 3
        },
        {
          "alias": "",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "hide": false,
          "refId": "A",
          "scenarioId": "random_walk",
          "seriesCount": 3
        }
      ],
      "thresholds": [],
      "timeRegions": [],
      "title": "Graph old (queries with multiple series)",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "mode": "time",
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "logBase": 1,
          "show": true
        },
        {
          "format": "short",
          "logBase": 1,
          "show": true
        }
      ],
      "yaxis": {
        "align": false
      }
    }
  ],
  "schemaVersion": 33,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "41303 stable query response ordering",
  "uid": "6CE3m9cnk",
  "version": 1,
  "weekStart": ""
}
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -268,3 +268,5 @@ replace gopkg.in/macaron.v1 => ./pkg/macaron
 replace github.com/go-macaron/binding => ./pkg/macaron/binding
 
 replace github.com/hashicorp/consul => github.com/hashicorp/consul v1.10.2
+
+replace github.com/grafana/grafana-plugin-sdk-go => github.com/grafana/grafana-plugin-sdk-go v0.116.1-0.20211118100855-44332cc5651c
```
Comment (sunker) [inline go.mod:272]: just a reminder to replace this

**Review by sunker — APPROVED** [review]
Tested this with two panels, one for CloudWatch and one for Prometheus, that I know randomly failed before. Cannot reproduce the problem with this fix. :+1: 

>Not sure this fixes all problems for CloudWatch - might need client side changes as well perhaps, see below. I've only tested with multiple CloudWatch queries that returns single series and that seems to work as expected. Before changes testing this I could see color changing between series, but haven't been able to reproduce this with these changes at least

I'm not 100% sure either, but I haven't seen any issues now that I've been testing this. Looking at the issues that you've linked to, they all seem to be related to ref ids and not the order of timeseries for a given query. Also, the cloudwatch frontend doesn't perform any custom manipulation to the response. https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/cloudwatch/datasource.ts#L464-L474

**Comment by mckn** [comment]
@marefr sounds like this is another one related to this issue: https://github.com/grafana/grafana/issues/41557

**Comment by marefr** [comment]
Backported via #42184

**Pull request merged**